src/Features/Core/Portable/ConvertForToForEach/AbstractConvertForToForEachCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,427 @@
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.ConvertForToForEach
+{
+    internal abstract class AbstractConvertForToForEachCodeRefactoringProvider<
+        TStatementSyntax,
+        TForStatementSyntax,
+        TExpressionSyntax,
+        TMemberAccessExpressionSyntax,
+        TTypeNode,
+        TVariableDeclaratorSyntax> : CodeRefactoringProvider
+        where TStatementSyntax : SyntaxNode
+        where TForStatementSyntax : TStatementSyntax
+        where TExpressionSyntax : SyntaxNode
+        where TMemberAccessExpressionSyntax : SyntaxNode
+        where TTypeNode : SyntaxNode
+        where TVariableDeclaratorSyntax : SyntaxNode
+    {
+        protected abstract string GetTitle();
+
+        protected abstract SyntaxNode ConvertForNode(
+            TForStatementSyntax currentFor, TTypeNode typeNode, SyntaxToken foreachIdentifier, 
+            TExpressionSyntax collectionExpression, ITypeSymbol iterationVariableType, OptionSet options);
+        protected abstract SyntaxList<TStatementSyntax> GetBodyStatements(TForStatementSyntax forStatement);
+        protected abstract bool IsValidVariableDeclarator(TVariableDeclaratorSyntax firstVariable);
+        protected abstract bool TryGetForStatementComponents(
+            TForStatementSyntax forStatement,
+            out SyntaxToken iterationVariable, out TExpressionSyntax initializer,
+            out TMemberAccessExpressionSyntax memberAccess, out TExpressionSyntax stepValue,
+            CancellationToken cancellationToken);
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var cancellationToken = context.CancellationToken;
+            var document = context.Document;
+
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var token = root.FindToken(context.Span.Start);
+
+            // position has to be inside the 'for' span, or if there is a selection, it must
+            // match the 'for' span exactly.
+            if (context.Span.IsEmpty && !token.Span.IntersectsWith(context.Span.Start))
+            {
+                return;
+            }
+
+            if (!context.Span.IsEmpty && context.Span != token.Span)
+            {
+                return;
+            }
+
+            var forStatement = token.Parent.AncestorsAndSelf().OfType<TForStatementSyntax>().FirstOrDefault();
```

Comment:
this is a little weird. isn't there something like FirstAncestorOrSelf<> or GetAncestorOrSelf<> you could use?

Reply:
yup. #resolved